### DISCUSSION TITLE
Disable clang diagnostic errors from headers

### DISCUSSION THREAD

**Discussion by H-G-Hristov** [microsoft/vscode-cpptools#9710] [Q&A]
Can I disable clang-diagnostic errors from headers as seen on the screenshot. This might be actually a bug as my `#include "System.hpp"` doesn't include any of these files which are included in bellow.

The system is macOS.

![Screenshot 2022-08-10 at 19 03 06](https://user-images.githubusercontent.com/47526411/183958328-7334b745-ef46-4d66-926f-dcecd87876e3.png)

Another Issue that might be a bug (this header is obviously not main):

![Screenshot 2022-08-10 at 19 07 34](https://user-images.githubusercontent.com/47526411/183958925-32c16eeb-71f6-4426-aa30-e7ad60817d54.png)

**Comment by sean-mcmanus** [comment]
Adding a setting like
```json
        "C_Cpp.codeAnalysis.exclude": {
            "**/utilities/include": true,
       }
```
should work.

The `#pragma once` error can occur if we're able to detect a source file that includes the header, in which case code analysis would be run on the header as a stand alone file.

**Comment by Zingam** [comment]
Is it possible to just have a on/off switch to exclude all system headers and/or third party headers, for example anything not in ${CMAKE_SORUCE_DIR}, without getting into complex regex, etc?

I'm sorry if this sounds stupid. I'm ignorant of this area.

**Reply by sean-mcmanus** [reply]
@Zingam Anything not in the ${workpspaceFolder} should already be omitted/filtered via the --header-filter regex which matches the workspace folder.

**Comment by rababR1** [comment]
I get the same issue but on Windows:
![image](https://github.com/microsoft/vscode-cpptools/assets/65360255/6ce81360-00bf-4fea-b6a7-61afc32edcdb)

**Reply by abdoei** [reply]
What is the setting rule you appended to the `settings.json` file? It should be like this 
```json
"C_Cpp.codeAnalysis.exclude": {
            "**/include": true,
    }
```
replace `**` with any number of directory levels above your compiler's `include` directory to prevent it from excluding your `${workspaceFolder}/include` from being analysed. Mine is `/usr/include/c++/14.2.1/` so if it is making trouble for me I would make `"**/include": true,` to `"**usr/include": true,` or for this specific case it would be `"/usr/include": true,`.

By the way, could you mark this discussion as solved if you are satisfied with the answers? @H-G-Hristov
